### DISCUSSION TITLE
Ability to define global @psalm-type in config

### DISCUSSION THREAD

**Discussion by M1ke** [vimeo/psalm#5376] [Ideas]
Using `@psalm-type` and `@psalm-import-type` is useful for places with a few uses of an object-like array. However when object-like arrays have been used to imply whole data strctures these imports may be used in hundreds of files and it's extra work to always import them.

Has it ever been considered for Psalm to define a "pre-import" list where a set of types can be imported that are considered available globally? This could conflict I guess with other named classes; my view is that if someone wants to use this, then they should maintain a sensible list and if Psalm finds two definitions it should error much like if two imports shared the same name without an alias.

I had wondered if a plugin could provide this - could I write a custom plugin to define a limited set of such data types for our use?

**Comment by weirdan** [comment]
> these imports may be used in hundreds of files and it's extra work to always import them.

`use` serves the same purpose (conflict prevention/resolution), and people don't seem to consider it too much extra work. Perhaps it's because all modern editors provide auto-import. Psalm could do this as well, through `psalm-language-server`. But somebody needs to actually implement it.

**Reply by M1ke** [reply]
Absolutely I think IDE support would change the game on that; however I don't imagine it would be a simple improvement on the face of it - that said my AST knowledge is very limited so maybe it is.

Given that the language server implementation would be hard, is there a way for a plugin to set types globally? That way an individual project could implement this relatively easily without it having to be a feature added to core.
